### PR TITLE
[versioning] EOS-24951: Sortable and globally unique version IDs

### PR DESCRIPTION
**Pull request by keithpine**
# Problem Statement

Generate version IDs for objects using an updated algorithm:

 1. Version ID prefix is a reverse timestamp in Base62 encoding
 2. Version ID suffix is 128-bits or more random data (UUID)

A Base62 encoding of reverse timestamps naturally sorts version IDs from newest to oldest for a lexicographical sorting order. Therefore the version ID can be used directly as part of a Motr index key.

The random UUID ensures that each version ID is globally unique for objects stored concurrently at the same millisecond, either within a cluster or between replicating clusters.

**NOTE: This change is not backward compatible. If this is not acceptable, I'll need to add in a change to detect which kind of version ID is being used and generate the object key appropriately.**

There are a few todos left in the PR:

1. Replace S3Uuid with a full 18-bytes of random data. Need to determine the best method for generating the random data.
2. Perform some benchmarking of the Base62 code. If there's no difference with Base64, consider using only Base62 (64 + 64).

# Design
- [Versioning HLD (Version ID section)](https://seagate-systems.atlassian.net/wiki/spaces/PUB/pages/562593803/Versioning#Version-IDs)
- [Version ID LLD](https://seagate-systems.atlassian.net/wiki/spaces/PUB/pages/784105532/Version+ID+LLD)

# Coding
   Checklist for Author
- [X] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
- [X] Unit and System Tests are added
- [X] Test Cases cover Happy Path, Non-Happy Path and Scalability
- [ ] Testing was performed with RPM

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
- [X] Interface change (if any) are documented
- [x] Side effects on other features (deployment/upgrade)
- [x] Dependencies on other component(s)

# Review Checklist 
  Checklist for Author
- [X] JIRA number/GitHub Issue added to PR
- [X] PR is self reviewed
- [x] Jira and state/status is updated and JIRA is updated with PR link
- [X] Check if the description is clear and explained

# Documentation
  Checklist for Author
- [x] Changes done to WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

**Comment by keithpine** [comment]
Pre-Merge Results: http://eos-jenkins.colo.seagate.com/job/S3server/job/S3Server-PreMerge-Test/821/

_Currently Running_

File: ut/s3_object_metadata_test.cc [hunk 1]
```diff
@@ -342,7 +342,8 @@ TEST_F(S3ObjectMetadataTest, LoadObjectInfoFromVersionListIdex) {
   std::string obj_ver_id = "MTg0NDY3NDI0MzcyNjA2NzM4Mj";
   metadata_obj_under_test->set_version_id(obj_ver_id);
   EXPECT_CALL(*(motr_kvs_reader_factory->mock_motr_kvs_reader),
-              get_keyval(_, "objectname/1844674243726067382", _, _)).Times(1);
+              get_keyval(_, "objectname/MTg0NDY3NDI0MzcyNjA2NzM4Mj", _, _))
+      .Times(1);
```
Comment (keithpine) [inline ut/s3_object_metadata_test.cc:346]: This illustrates where this change is no longer backward compatible. If this is required, then this change would be reverted and I'd need to detect which ID is being used.

**Review by ajay-paratmandali — APPROVED** [review]

**Review by sanalpk — APPROVED** [review]

**Comment by indrajitzagade** [comment]
@keithpine I tried resolving the merge conflicts here. Please check if it was fine. Thanks.

**Review by sjain09 — APPROVED** [review]

**Comment by keithpine** [comment]
@indrajitzagade There were a few issues with the merge. I rebased to the latest and it's fine now.

Pre-merge test result, passing: http://eos-jenkins.colo.seagate.com/job/S3server/job/S3Server-PreMerge-Test/926/

If there are any later concerns about backwards compatibility questions I raised, it can be addressed in a follow up PR then.

**Pull request merged**